aptos-move/framework/aptos-framework/sources/configs/Stake.move

```diff
@@ -276,167 +413,313 @@ module AptosFramework::Stake {
     }
 
-    /// Mint the reward and add to the delegation based on some formula
-    fun distribute_reward(v: &mut IterableTable<address, Delegation>) {
-        let key = IterableTable::head_key(v);
-        while (Option::is_some(&key)) {
-            let (delegation, _, next) = IterableTable::borrow_iter_mut(v, Option::borrow(&key));
-            let reward = Coin::zero<TestCoin>(); // mint some coins based on delegation, timestamp, maybe also total stakes
-            Coin::merge<TestCoin>(&mut delegation.rewards, reward);
-            key = next;
+    /// For each delegator, withdraw coins from `pending_pool` and deposit in `pool`.
+    fun process_pending_pool(
+        pool: &mut CoinPool,
+        pending_pool: &mut PendingCoinPool,
+        new_delegation_state: u64,
+    ) acquires Delegation {
+        let delegator_addresses = &mut pending_pool.delegator_addresses;
+        let i = 0;
+        let len = Vector::length(delegator_addresses);
+        while (i < len) {
+            let delegator_address = *Vector::borrow(delegator_addresses, i);
+            let delegation = borrow_global_mut<Delegation>(delegator_address);
+            let coins = extract_from_pool(&mut pending_pool.pool, delegation.shares);
+            delegation.shares = buy_shares_from_pool(pool, coins);
+            delegation.state = new_delegation_state;
+
+            i = i + 1;
         };
+        // There might be some coin residual (dust) left in the pending_pool but this should not affect operations.
+        pending_pool.pool.total_shares = 0;
+        pending_pool.delegator_addresses = Vector::empty();
     }
 
-    fun append<T>(v1: &mut vector<T>, v2: &mut vector<T>) {
-        while (!Vector::is_empty(v2)) {
-            Vector::push_back(v1, Vector::pop_back(v2));
+    /// Distribute reward to each delegation.
+    fun distribute_reward(pool: &mut CoinPool) {
+        let rewards = Coin::zero<TestCoin>();
+        Coin::merge<TestCoin>(&mut pool.coins, rewards);
+    }
+
+    /// Buy the shares in `pool` with `coins`.
+    fun buy_shares_from_pool(pool: &mut CoinPool, coins: Coin<TestCoin>): u64 {
+        let buy_in_amount = Coin::value<TestCoin>(&coins);
+        let pool_balance = Coin::value<TestCoin>(&pool.coins);
+        Coin::merge<TestCoin>(&mut pool.coins, coins);
+
+        // If pool has no balance yet, issue as many shares as the buy in amount of coins.
+        // This is effectively issuing shares at 1:1 value to coin.
+        if (pool_balance == 0) {
+            pool.total_shares = buy_in_amount;
+            buy_in_amount
+        } else {
+            // Multiply first to avoid precision loss from integer division.
+            // u64 can hold up to ~10**19 so this shouldn't overflow if maximum_stake is below 10**9 (1B).
+            // total_shares and buy_in_amount are at most maximum_stakes, so total_shares * buy_in_amount <= 10**18.
+            let shares_amount = buy_in_amount * pool.total_shares / pool_balance;
```

Comment:
I think shares should be a resource type (coin) to benefit from compiler checks

Reply:
That might have some tax implications though? Currently shares are just an internal accounting detail. If we start issuing an actual Coin<DelegationShares>, those shares would be transferrable.